src/main/java/com/google/api/gax/bundling/PushingBundler.java

```diff
@@ -0,0 +1,234 @@
+    this.executor = Preconditions.checkNotNull(builder.executor);
+    this.maxDelay = Preconditions.checkNotNull(builder.maxDelay);
+    this.receiver = Preconditions.checkNotNull(builder.receiver);
+    this.flowController = Preconditions.checkNotNull(builder.flowController);
+    this.bundleMerger = Preconditions.checkNotNull(builder.bundleMerger);
+
+    resetThresholds();
+  }
+
+  public static class Builder<E> {
+    private Collection<BundlingThreshold<E>> thresholds;
+    private ScheduledExecutorService executor;
+    private Duration maxDelay;
+    private ThresholdBundleReceiver<E> receiver;
+    private BundlingFlowController<E> flowController;
+    private BundleMerger<E> bundleMerger;
+
+    private Builder() {}
+
+    public Builder<E> setThresholds(Collection<BundlingThreshold<E>> thresholds) {
+      this.thresholds = thresholds;
+      return this;
+    }
+
+    public Builder<E> setExecutor(ScheduledExecutorService executor) {
+      this.executor = executor;
+      return this;
+    }
+
+    public Builder<E> setMaxDelay(Duration maxDelay) {
+      this.maxDelay = maxDelay;
+      return this;
+    }
+
+    public Builder<E> setReceiver(ThresholdBundleReceiver<E> receiver) {
+      this.receiver = receiver;
+      return this;
+    }
+
+    public Builder<E> setFlowController(BundlingFlowController<E> flowController) {
+      this.flowController = flowController;
+      return this;
+    }
+
+    public Builder<E> setBundleMerger(BundleMerger<E> bundleMerger) {
+      this.bundleMerger = bundleMerger;
+      return this;
+    }
+
+    public PushingBundler<E> build() {
+      return new PushingBundler<>(this);
+    }
+  }
+
+  public static <E> Builder<E> newBuilder() {
+    return new Builder<>();
+  }
+
+  /**
+   * Immediately send contained elements to the {@code ThresholdBundleReceiver}.
+   */
+  public void flush() {
+    try {
+      // We wait on the executor so that any running processing
+      executorProcess(removeBundle()).get();
+    } catch (InterruptedException | ExecutionException e) {
+    }
+  }
+
+  /**
+   * Adds an element to the bundler. If the element causes the collection to go past any of the
+   * thresholds, the bundle will be sent to the {@code ThresholdBundleReceiver}.
+   */
+  public void add(E e) throws FlowControlException {
+    // We need to reserve resources from flowController outside the lock, so that they can be
+    // released by process().
+    flowController.reserve(e);
+    lock.lock();
+    try {
+      receiver.validateBundle(e);
+      if (currentOpenBundle == null) {
+        currentOpenBundle = e;
+        currentAlarmFuture =
+            executor.schedule(flushRunnable, maxDelay.getMillis(), TimeUnit.MILLISECONDS);
+      } else {
+        bundleMerger.merge(currentOpenBundle, e);
+      }
+
+      if (isAnyThresholdReached(e)) {
+        // We must use the executor here because this could be a client thread on a hot path
+        executorProcess(removeBundle());
```

Comment:
This makes sense. I had initially created a dedicated executor, but an async receiver is a good approach I think.